runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/jobsubmission/JobService.java

```diff
@@ -0,0 +1,309 @@
+  public void getMessageStream(
+      JobMessagesRequest request,
+      StreamObserver<JobMessagesResponse> responseObserver) {
+    try {
+      String invocationId = request.getJobId();
+      JobInvocation invocation = invocations.get(invocationId);
+
+      Function<JobState.Enum, JobMessagesResponse> stateResponseFunction =
+          state ->
+              JobMessagesResponse
+                  .newBuilder()
+                  .setStateResponse(GetJobStateResponse.newBuilder().setState(state).build())
+                  .build();
+      TransformStreamObserver<JobState.Enum, JobMessagesResponse> stateObserver =
+          TransformStreamObserver.create(stateResponseFunction, responseObserver);
+
+      Function<JobMessage, JobMessagesResponse> messagesResponseFunction =
+          message -> JobMessagesResponse.newBuilder().setMessageResponse(message).build();
+      TransformStreamObserver<JobMessage, JobMessagesResponse> messageObserver =
+          TransformStreamObserver.create(messagesResponseFunction, responseObserver);
+
+      synchronized (invocation) {
+        invocation.addStateObserver(stateObserver);
+        invocation.addMessageObserver(messageObserver);
+      }
+    } catch (Exception e) {
+      LOG.error("Encountered Unexpected Exception", e);
+      responseObserver.onError(Status.INTERNAL.withCause(e).asException());
+    }
+  }
+
+  @Override
+  public void close() throws Exception {
+    for (JobPreparation preparation : ImmutableList.copyOf(preparations.values())) {
+      try {
+        preparation.stagingService().close();
+      } catch (Exception e) {
+        LOG.warn("Exception while closing job {}", preparation);
+      }
+    }
+  }
+
+  /**
+   * Utility class that transforms observations into one that an output observer can accept.
+   *
+   * <p>This class synchronizes on outputObserver when handling it, allowing for multiple
+   * TransformStreamObservers to multiplex to a single outputObserver.
+   *
+   * @param <T1> Input Type
+   * @param <T2> Output Type
+   */
+  private static class TransformStreamObserver<T1, T2> implements StreamObserver<T1> {
+
+    /**
+     * Create a new TransformStreamObserver.
+     *
+     * @param transform The function used to transform observations.
+     * @param outputObserver The observer to forward transform outputs to.
+     */
+    public static <T1, T2> TransformStreamObserver<T1, T2> create(
+        Function<T1, T2> transform, StreamObserver<T2> outputObserver) {
+      return new TransformStreamObserver<>(transform, outputObserver);
+    }
+
+    private final Function<T1, T2> transform;
+    private final StreamObserver<T2> outputObserver;
+
+    private TransformStreamObserver(Function<T1, T2> transform, StreamObserver<T2> outputObserver) {
+      this.transform = transform;
+      this.outputObserver = outputObserver;
+    }
+
+    @Override
+    public void onNext(T1 i) {
+      T2 o = transform.apply(i);
+      synchronized (outputObserver) {
```

Comment:
Here too, thread safety should be responsibility of outputObserver itself.

Reply:
It was my understanding that individual stream observers are explicitly *not* threadsafe:
https://grpc.io/grpc-java/javadoc/io/grpc/stub/StreamObserver.html

Since in this case multiple TransformStreamObservers can be multiplexed onto a single internal StreamObserver, I thought this would be a good place to synchronize.  Am I missing something?  I guess I could explicitly create another threadsafe StreamObserver that this composes, but I feel like it would be a bit overkill for a private class.

Reply:
This class will get replaced with a Consumer<> implementation, but it will compose a StreamObserver instance so I think the question is still relevant.